db/changes.go

```diff
@@ -820,3 +820,106 @@ func (db *Database) getLateFeed(feedHandler *lateSequenceFeed) (<-chan *ChangeEn
 func (db *Database) closeLateFeed(feedHandler *lateSequenceFeed) {
 	db.changeCache.getChannelCache(feedHandler.channelName).ReleaseLateSequenceClient(feedHandler.lastSequence)
 }
+
+/*
+ * Generate the changes for a specific list of doc ID's, only documents accesible to the user will generate
+ * results.  Only supports non-continuous changes, closes buffered channel before returning.
+ */
+func (db *Database) DocIDChangesFeed(userChannels base.Set, explicitDocIds []string, options ChangesOptions) (<-chan *ChangeEntry, error) {
+
+	// Subroutine that creates a response row for a document:
+	output := make(chan *ChangeEntry, len(explicitDocIds))
+	rowMap := make(map[uint64]*ChangeEntry)
+
+	createChangeEntry := func(docid string) *ChangeEntry {
+		row := &ChangeEntry{ID: docid}
+
+		// Fetch the document body and other metadata that lives with it:
+		populatedDoc, body, err := db.GetDocAndActiveRev(docid)
+		if err != nil {
+			base.LogTo("Changes", "Unable to get changes for docID %v, caused by %v", docid, err)
+			return nil
+		}
+
+		if populatedDoc.Sequence <= options.Since.Seq {
+			return nil
+		}
+
+		if body == nil {
+			return nil
+		}
+
+		changes := make([]ChangeRev, 1)
+		changes[0] = ChangeRev{"rev": body["_rev"].(string)}
+		row.Changes = changes
+		row.Seq = SequenceID{Seq: populatedDoc.Sequence}
+		row.SetBranched((populatedDoc.Flags & channels.Branched) != 0)
+
+		var removedChannels []string
+
+		if deleted, _ := body["_deleted"].(bool); deleted {
+			row.Deleted = true
+		}
+
+		userCanSeeDocChannel := false
+
+		if db.user == nil || db.user.Channels().Contains(channels.UserStarChannel) {
+			userCanSeeDocChannel = true
+		} else if len(populatedDoc.Channels) > 0 {
+			//Do special _removed/_deleted processing
+			for channel, removal := range populatedDoc.Channels {
+				//Doc is tagged with channel or was removed at a sequence later that since sequence
+				if removal == nil || removal.Seq > options.Since.Seq {
+					//if the current user has access to this channel
+					if db.user.CanSeeChannel(channel) {
+						userCanSeeDocChannel = true
+						//If the doc has been removed
+						if removal != nil {
+							removedChannels = append(removedChannels, channel)
+							if removal.Deleted {
+								row.Deleted = true
+							}
+						}
+					}
+				}
+			}
+		}
```

Comment:
- Can you add comment for this block of code (line 866->886) to describe the overall functionality?
- I (subjectively) feel this block has "too many" levels of indentation.  Is it feasible to reduce this? 
- I feel like refactoring out a function/method for this block would make it a bit easier to grok the overall behavior of `createChangeEntry()`

Reply:
Added some comments and pared down one level of indentation.  Filed a ticket for a post-2.0 unit test that could include some refactoring of this method as well.